File: frontend/src/api/analytics.ts
import { api } from './client';
import type {
  ClientKind,
  ClientStatus,
  EventStatus,
  VacancyStatus,
} from './types';

export type CompareMode = 'prev' | 'yoy' | 'none';
export type Granularity = 'auto' | 'day' | 'week' | 'month';

export interface PeriodWindow {
  from: string;
  to: string;
}

export interface CompareWindow extends PeriodWindow {
  mode: CompareMode;
}

export interface DashboardSummary {
  openVacancies: number;
  activeCandidates: number;
  closedThisMonth: number;
  hiredThisMonth: number;
  delta: {
    openVacancies: number;
    activeCandidates: number;
    closedThisMonth: number;
    hiredThisMonth: number;
  };
  period: PeriodWindow;
  compare: CompareWindow | null;
}

export interface FunnelBucket {
  status: VacancyStatus;
  count: number;
}

export interface RecruiterLoad {
  recruiterId: string;
  activeCount: number;
}

export interface TrendsPoint {
  bucket: string;
  value: number;
}

export interface TrendsResponse {
  granularity: Exclude<Granularity, 'auto'>;
  period: PeriodWindow;
  series: {
    vacanciesCreated: TrendsPoint[];
    vacanciesClosed: TrendsPoint[];
    candidatesCreated: TrendsPoint[];
    hires: TrendsPoint[];
  };
}

export interface SummaryParams {
  from?: string;
  to?: string;
  compare?: CompareMode;
}

export interface TrendsParams {
  from?: string;
  to?: string;
  granularity?: Granularity;
}

export interface PeriodParams {
  from?: string;
  to?: string;
}

// ─── Funnel v2 ────────────────────────────────────────────────────────

export type MatchStatus =
  | 'submitted'
  | 'reviewed'
  | 'interview'
  | 'offered'
  | 'accepted'
  | 'rejected_client'
  | 'rejected_internal';

export interface FunnelStage {
  status: MatchStatus;
  count: number;
  conversionPct: number;
  dropOff: number;
}

export interface FunnelResponse {
  stages: FunnelStage[];
  rejected: { client: number; internal: number; total: number };
  total: number;
  overallConversionPct: number;
  period: PeriodWindow;
}

// ─── Time-to-hire ─────────────────────────────────────────────────────

export interface TimeToHireBucket {
  label: string;
  maxDays: number | null;
  count: number;
}

export interface TimeInStage {
  status: string;
  avgDays: number;
  medianDays: number;
  sample: number;
}

export interface TimeToHireResponse {
  sampleSize: number;
  avgDays: number;
  medianDays: number;
  p90Days: number;
  distribution: TimeToHireBucket[];
  byStage: TimeInStage[];
  period: PeriodWindow;
}

// ─── Attention ────────────────────────────────────────────────────────

export interface AttentionVacancyItem {
  id: string;
  title: string;
  status?: VacancyStatus | null;
  daysInStatus?: number | null;
  daysOpen?: number | null;
  deadline?: string | null;
  daysOverdue?: number | null;
  daysLeft?: number | null;
}

export interface AttentionCandidateItem {
  id: string;
  fullName: string;
  status: string;
  daysInStatus: number;
}

export interface AttentionVacancyBlock {
  total: number;
  thresholdDays?: number | null;
  items: AttentionVacancyItem[];
}

export interface AttentionCandidateBlock {
  total: number;
  thresholdDays?: number | null;
  items: AttentionCandidateItem[];
}

export interface AttentionResponse {
  stuckVacancies: AttentionVacancyBlock;
  stuckCandidates: AttentionCandidateBlock;
  vacanciesWithoutCandidates: AttentionVacancyBlock;
  overdueDeadlines: AttentionVacancyBlock;
  deadlinesNext7Days: AttentionVacancyBlock;
  deadlinesNext14Days: { total: number };
}

// ─── Recruiter performance ────────────────────────────────────────────

export interface RecruiterMetric {
  recruiterId: string;
  fullName: string;
  candidatesCreated: number;
  presented: number;
  hired: number;
  hireRatePct: number;
  avgTimeToHireDays: number;
  totalMargin: number;
  sparkline: number[];
}

export interface RecruiterPerformanceResponse {
  items: RecruiterMetric[];
  period: PeriodWindow;
}

// ─── Client performance ──────────────────────────────────────────────

export type ClientHealthFlag =
  | 'stale'
  | 'no_open'
  | 'high_rejection'
  | 'no_vacancies_ever';

export interface ClientMetric {
  clientId: string;
  name: string;
  industry: string;
  status?: ClientStatus | null;
  clientKind?: ClientKind | null;
  vacanciesTotal: number;
  vacanciesOpen: number;
  vacanciesClosedInPeriod: number;
  hiresInPeriod: number;
  avgTimeToFillDays: number;
  monthlyMarginRunRate: number;
  presentedToHiredPct: number;
  lastVacancyAt?: string | null;
  daysSinceLastVacancy?: number | null;
  rejectionRatePct: number;
  healthFlags: ClientHealthFlag[];
  sparkline: number[];
}

export interface ClientPerformanceResponse {
  items: ClientMetric[];
  period: PeriodWindow;
}

// ─── Weekly activity («Итоги недели») ─────────────────────────────────

export interface WeeklyVacancyItem {
  id: string;
  title: string;
  status: VacancyStatus;
  createdAt: string;
  clientId: string;
  clientName: string;
}

export interface WeeklyVacancyBlock {
  total: number;
  items: WeeklyVacancyItem[];
}

export interface WeeklySubmissionItem {
  matchId: string;
  status: MatchStatus;
  addedAt: string;
  candidateId: string;
  candidateName: string;
  vacancyId: string;
  vacancyTitle: string;
  clientName: string;
  addedByName?: string | null;
}

export interface WeeklySubmissionBlock {
  total: number;
  items: WeeklySubmissionItem[];
}

/** Собеседование, назначенное на окно (startsAt внутри периода). */
export interface WeeklyInterviewItem {
  eventId: string;
  title: string;
  startsAt: string;
  status: EventStatus;
  candidateId?: string | null;
  candidateName?: string | null;
  vacancyId?: string | null;
  vacancyTitle?: string | null;
}

export interface WeeklyInterviewBlock {
  total: number;
  items: WeeklyInterviewItem[];
}

/** Строка разбивки по сотруднику; userId/fullName = null — «не указан». */
export interface WeeklyUserCount {
  userId?: string | null;
  fullName?: string | null;
  count: number;
}

export interface WeeklyActivityResponse {
  period: PeriodWindow;
  newVacancies: WeeklyVacancyBlock;
  submittedCandidates: WeeklySubmissionBlock;
  /** Собеседования, назначенные на окно (canceled не считаются). */
  interviews: WeeklyInterviewBlock;
  /** Созданные вакансии по аккаунт-менеджерам. */
  byManagers: WeeklyUserCount[];
  /** Подачи по тем, кто прикрепил кандидата. */
  byRecruiters: WeeklyUserCount[];
}

// ─── Worklog (учёт времени в системе) ─────────────────────────────────

export type WorkSessionEndReason =
  | 'disconnect'
  | 'sweep'
  | 'server_shutdown'
  | 'reconcile';

export interface WorklogUserSummary {
  userId: string;
  fullName: string;
  totalSeconds: number;
  /** Активное время (вкладка видима + взаимодействие), пропорц. окну. */
  totalActiveSeconds: number;
  sessionsCount: number;
}

export interface WorklogSummaryResponse {
  from: string;
  to: string;
  items: WorklogUserSummary[];
}

export interface WorklogSession {
  id: string;
  userId: string | null;
  startedAt: string;
  lastHeartbeatAt: string;
  endedAt: string | null;
  endReason: WorkSessionEndReason | null;
  durationSeconds: number | null;
}

export interface WorklogParams {
  from?: string;
  to?: string;
  /** Конкретный сотрудник; для непривилегированных ролей игнорируется бэком. */
  userId?: string;
}

function toSearchParams(obj: Record<string, unknown>): URLSearchParams {
  const p = new URLSearchParams();
  for (const [k, v] of Object.entries(obj)) {
    if (v !== undefined && v !== null) p.set(k, String(v));
  }
  return p;
}

export const analyticsApi = {
  summary: (params: SummaryParams = {}) =>
    api
      .get('analytics/summary', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<DashboardSummary>(),
  funnel: () => api.get('analytics/funnel').json<FunnelBucket[]>(),
  recruiterLoad: () => api.get('analytics/recruiter-load').json<RecruiterLoad[]>(),
  trends: (params: TrendsParams = {}) =>
    api
      .get('analytics/trends', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<TrendsResponse>(),
  funnelV2: (params: PeriodParams = {}) =>
    api
      .get('analytics/funnel-v2', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<FunnelResponse>(),
  timeToHire: (params: PeriodParams = {}) =>
    api
      .get('analytics/time-to-hire', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<TimeToHireResponse>(),
  attention: (top = 5) =>
    api
      .get('analytics/attention', {
        searchParams: toSearchParams({ top: String(top) }),
      })
      .json<AttentionResponse>(),
  recruiterPerformance: (params: PeriodParams = {}) =>
    api
      .get('analytics/recruiter-performance', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<RecruiterPerformanceResponse>(),
  clientPerformance: (params: PeriodParams = {}) =>
    api
      .get('analytics/client-performance', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<ClientPerformanceResponse>(),
  weeklyActivity: (params: PeriodParams = {}) =>
    api
      .get('analytics/weekly-activity', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<WeeklyActivityResponse>(),
  worklogSummary: (params: WorklogParams = {}) =>
    api
      .get('analytics/worklog/summary', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<WorklogSummaryResponse>(),
  worklogSessions: (params: WorklogParams = {}) =>
    api
      .get('analytics/worklog/sessions', {
        searchParams: toSearchParams(params as Record<string, unknown>),
      })
      .json<WorklogSession[]>(),
};

File: backend/app/api/v1/endpoints/analytics.py
"""Эндпоинты /analytics/*.

Доступ — авторизованный пользователь. Контентовая фильтрация по ролям
(account_manager видит только свои) появится в Этапе 8 (см. план §4).
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.errors import ApiError
from app.db.session import get_db
from app.modules.analytics import service, worklog_service
from app.modules.analytics.schemas import (
    AttentionResponse,
    ChatStats,
    ClientPerformanceResponse,
    DashboardSummary,
    FunnelBucket,
    FunnelResponse,
    RecruiterLoad,
    RecruiterPerformanceResponse,
    TimeToHireResponse,
    TrendsResponse,
    WeeklyActivityResponse,
)
from app.modules.analytics.worklog_schemas import (
    WorklogSession,
    WorklogSummaryResponse,
    WorklogUserSummary,
)
from app.modules.chat.analytics import chat_stats
from app.modules.auth.dependencies import get_current_user
from app.modules.users.models import Role, User

# Учёт времени — раздел только для администраторов.
def _ensure_worklog_admin(current: User) -> None:
    if current.role != Role.admin:
        raise ApiError(
            403, "forbidden", "Учёт времени доступен только администраторам"
        )

router = APIRouter(prefix="/analytics", tags=["analytics"])


@router.get("/summary", response_model=DashboardSummary, summary="Сводка дашборда")
async def summary(
    from_dt: datetime | None = Query(
        None,
        alias="from",
        description="Начало периода (ISO-8601). По умолчанию — начало текущего месяца.",
    ),
    to_dt: datetime | None = Query(
        None, alias="to", description="Конец периода (ISO-8601). По умолчанию — сейчас."
    ),
    compare: Literal["prev", "yoy", "none"] = Query(
        "prev",
        description="Окно для расчёта delta: prev (того же размера непосредственно перед), yoy (тот же диапазон год назад), none (без дельт).",
    ),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> DashboardSummary:
    period = service.resolve_period(from_dt, to_dt)
    data = await service.summary(db, period=period, compare_mode=compare)
    return DashboardSummary.model_validate(data)


@router.get("/funnel", response_model=list[FunnelBucket], summary="Воронка вакансий")
async def funnel(
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[FunnelBucket]:
    rows = await service.funnel(db)
    return [FunnelBucket.model_validate(r) for r in rows]


@router.get(
    "/recruiter-load", response_model=list[RecruiterLoad], summary="Нагрузка рекрутеров"
)
async def recruiter_load(
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[RecruiterLoad]:
    rows = await service.recruiter_load(db)
    return [RecruiterLoad.model_validate(r) for r in rows]


@router.get(
    "/trends",
    response_model=TrendsResponse,
    summary="Временные ряды: создано/закрыто вакансий, заведено кандидатов, наймы",
)
async def trends(
    from_dt: datetime | None = Query(None, alias="from"),
    to_dt: datetime | None = Query(None, alias="to"),
    granularity: Literal["auto", "day", "week", "month"] = Query(
        "auto",
        description="Гранулярность бакетов. auto: до 31 дня → day, до 180 → week, иначе → month.",
    ),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TrendsResponse:
    period = service.resolve_period(from_dt, to_dt)
    data = await service.trends(db, period=period, granularity=granularity)
    return TrendsResponse.model_validate(data)


@router.get(
    "/funnel-v2",
    response_model=FunnelResponse,
    summary="Воронка по VacancyCandidate с конверсиями и drop-off",
)
async def funnel_v2(
    from_dt: datetime | None = Query(None, alias="from"),
    to_dt: datetime | None = Query(None, alias="to"),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> FunnelResponse:
    period = service.resolve_period(from_dt, to_dt)
    data = await service.funnel_v2(db, period=period)
    return FunnelResponse.model_validate(data)


@router.get(
    "/time-to-hire",
    response_model=TimeToHireResponse,
    summary="Скорость найма: avg/median/p90, распределение, время по стадиям",
)
async def time_to_hire(
    from_dt: datetime | None = Query(None, alias="from"),
    to_dt: datetime | None = Query(None, alias="to"),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TimeToHireResponse:
    period = service.resolve_period(from_dt, to_dt)
    data = await service.time_to_hire(db, period=period)
    return TimeToHireResponse.model_validate(data)


@router.get(
    "/attention",
    response_model=AttentionResponse,
    summary="Требует внимания: зависшие вакансии/кандидаты, дедлайны",
)
async def attention(
    top: int = Query(5, ge=1, le=20),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AttentionResponse:
    data = await service.attention(db, top=top)
    return AttentionResponse.model_validate(data)


@router.get(
    "/recruiter-performance",
    response_model=RecruiterPerformanceResponse,
    summary="Эффективность рекрутеров за период",
)
async def recruiter_performance(
    from_dt: datetime | None = Query(None, alias="from"),
    to_dt: datetime | None = Query(None, alias="to"),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> RecruiterPerformanceResponse:
    period = service.resolve_period(from_dt, to_dt)
    data = await service.recruiter_performance(db, period=period)
    return RecruiterPerformanceResponse.model_validate(data)


@router.get(
    "/client-performance",
    response_model=ClientPerformanceResponse,
    summary="Аналитика по клиентам за период",
)
async def client_performance(
    from_dt: datetime | None = Query(None, alias="from"),
    to_dt: datetime | None = Query(None, alias="to"),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ClientPerformanceResponse:
    period = service.resolve_period(from_dt, to_dt)
    data = await service.client_performance(db, period=period)
    return ClientPerformanceResponse.model_validate(data)


@router.get(
    "/weekly-activity",
    response_model=WeeklyActivityResponse,
    summary="Итоги недели: новые вакансии и поданные кандидаты за окно",
)
async def weekly_activity(
    from_dt: datetime | None = Query(
        None,
        alias="from",
        description="Начало окна (ISO-8601). По умолчанию — понедельник текущей недели, 00:00 UTC.",
    ),
    to_dt: datetime | None = Query(None, alias="to"),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> WeeklyActivityResponse:
    if from_dt is None and to_dt is None:
        period = service.week_period()
    else:
        period = service.resolve_period(from_dt, to_dt)
    data = await service.weekly_activity(db, period=period)
    return WeeklyActivityResponse.model_validate(data)


@router.get(
    "/chat",
    response_model=ChatStats,
    summary="Метрики чата (Этап 6 чата)",
)
async def chat_metrics(
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ChatStats:
    data = await chat_stats(db)
    return ChatStats.model_validate(data)


# === Учёт времени сотрудников (work_sessions) — только администраторы =======


@router.get(
    "/worklog/summary",
    response_model=WorklogSummaryResponse,
    summary="Учёт времени: суммарное время по сотрудникам за период (admin)",
)
async def worklog_summary(
    from_dt: datetime | None = Query(None, alias="from"),
    to_dt: datetime | None = Query(None, alias="to"),
    user_id: uuid.UUID | None = Query(None, alias="userId"),
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> WorklogSummaryResponse:
    _ensure_worklog_admin(current)
    period = service.resolve_period(from_dt, to_dt)
    user_ids = None if user_id is None else [user_id]
    rows = await worklog_service.summary(
        db, from_dt=period.from_dt, to_dt=period.to_dt, user_ids=user_ids
    )
    return WorklogSummaryResponse(
        from_dt=period.from_dt,
        to_dt=period.to_dt,
        items=[WorklogUserSummary.model_validate(r) for r in rows],
    )


@router.get(
    "/worklog/sessions",
    response_model=list[WorklogSession],
    summary="Учёт времени: сырые интервалы пользователя за период (admin)",
)
async def worklog_sessions(
    from_dt: datetime | None = Query(None, alias="from"),
    to_dt: datetime | None = Query(None, alias="to"),
    user_id: uuid.UUID | None = Query(None, alias="userId"),
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[WorklogSession]:
    _ensure_worklog_admin(current)
    period = service.resolve_period(from_dt, to_dt)
    # Без явного userId админ смотрит свои интервалы (список «по всем» бессмыслен).
    target = user_id or current.id
    rows = await worklog_service.sessions(
        db, user_id=target, from_dt=period.from_dt, to_dt=period.to_dt
    )
    return [WorklogSession.model_validate(r) for r in rows]
